Additional autograd unit tests for Python UDFs.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28824 Additional autograd unit tests for Python UDFs.**

1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

Commit:
Additional autograd unit tests for Python UDFs.

1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -69,6 +69,23 @@ struct python_error : public std::exception {
+    // Retrieve the error message from the value.
+    if (value != nullptr && value != Py_None) {
+      AutoGIL gil;
+      Py_XINCREF(value);
```
Manual refcounting, blegh

ezyang replied:
I don't see a paired decref, nor do I see a steal on value, so it looks to me like you're leaking `value`?

ezyang commented on torch/csrc/autograd/python_engine.h:
```diff
@@ -21,6 +21,10 @@ struct PythonEngine : public Engine {
+
+  variable_list execute_with_graph_task(
+      std::shared_ptr<GraphTask> graph_task,
+      std::shared_ptr<Node> graph_root);
```
Mark this `override` right?

ezyang commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -70,6 +70,16 @@ variable_list PythonEngine::execute(
+  try {
+    return Engine::execute_with_graph_task(graph_task, graph_root);
+  } catch (python_error& e) {
+    e.restore();
```
This looks pretty weird to me. What it kind of looks like is that you reraised a `python_error` from another thread but forgot to restore the Python error context at the time you threw the error, so now you have to "fix" it here. Wouldn't it be better to just handle it properly at the rethrowing site?

pritamdamania87 replied:
There is a case where this method is called directly from `DistEngine` in that case, we would like to restore the error. I'll change this to only do `e.restore` if the error is not set already.

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -69,6 +69,23 @@ struct python_error : public std::exception {
+      AutoGIL gil;
+      Py_XINCREF(value);
+      PyObject* pyStr = PyObject_Str(value);
+      if (pyStr != Py_None && pyStr != nullptr) {
```
Even if `pyStr` returns a `Py_None` you are obligated to decref it (yes, `Py_None` is refcounted)

ezyang replied:
If it returns `nullptr` you are obligated to propagate the Python exception, or clear it

ezyang commented on torch/csrc/Exceptions.h:
```diff
@@ -69,6 +69,23 @@ struct python_error : public std::exception {
+      PyObject* pyStr = PyObject_Str(value);
+      if (pyStr != Py_None && pyStr != nullptr) {
+        PyObject* str = PyUnicode_AsEncodedString(pyStr, "utf-8", "strict");
+        const auto& ret = PyBytes_AS_STRING(str);
```
You can't do this. You need to keep `str` live for the returned `char*` to be valid, but there's no way in this API to do so. Your `char*` goes immediately to dead data after you do this

ezyang reviewed (changes requested):
Major errors

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
+                dist_autograd.backward([loss.sum()])
+
+
+    _backward_done = False
```
Other state vars are out of the test class. Any reason for putting this here?

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
+    def _wait_backward_done():
+        while not DistAutogradTest._backward_done:
+            time.sleep(0.1)
+            pass
```
why do we need this `pass` here?

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
+        t1 = torch.rand((3, 3), requires_grad=True)
+        t2 = torch.rand((3, 3), requires_grad=True)
+        with dist_autograd.context() as context_id:
+            loss = rpc.rpc_sync('worker{}'.format(self._next_rank()), DistAutogradTest._nested_rpc_call_backward_error, args=(t1, t2, self._next_rank()))
```
There are several super long lines. Let's break them into shorter ones.

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -784,6 +782,129 @@ def test_backward_different_dtypes(self):
+
+            if self.rank == 0:
+                # Wait a bit for rank 2 to die.
+                time.sleep(5)
```
This depends on physical time to sync processes, which might potentially make this test flaky. Is there a way to implement the ordering explicitly?

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

zhaojuanmao commented:
the tests look good to me, wondering whether it could be mixed with refs as well for these tests

ezyang reviewed (approved):
I didn't review the newly added tests carefully, please get someone else to review them

pritamdamania87 commented:
@zhaojuanmao I believe Shen added a bunch of tests for RRefs in https://github.com/pytorch/pytorch/pull/28656. Do you have any additional tests in mind?

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -103,6 +102,22 @@ def _all_contexts_cleaned_up(timeout_seconds=10):
+
+def wait_until_node_failure(rank):
+    '''
+        Loops until an RPC to the given rank doesn't fail. This is used to
```
you mean loops until ... DOES fail?

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Lint failures on `python_engine.cpp` does not seem relevant, but please double check. Tests look good to me.

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

Commit:
Update on "Additional autograd unit tests for Python UDFs."


1) Enhanced autograd unit tests to test the
torch.distributed.autograd.backward() API more thoroughly on Python UDFs.
2) Enhanced `python_error` to override `what` such that it returns an
appropriate error string if we call `what()` on this error. This ensures we can
propagate exceptions over the wire during RPCs (since we get the error string
by calling what() on the exception)

Differential Revision: [D18195584](https://our.internmc.facebook.com/intern/diff/D18195584/)

[ghstack-poisoned]

ezyang commented:
This diff caused C++ docs check to fail on master:

```
+ grep -v 'warning: Included by graph for '\''types.h'\'' not generated, too many nodes' doxygen-log.txt
+ set -e
+ mv temp.txt doxygen-log.txt
++ grep warning: doxygen-log.txt
++ wc -l
+ warnings=1
+ echo 'Treating all remaining warnings as errors'
+ [[ 1 -ne 0 ]]
+ echo 'Failing Doxygen test because the following warnings were treated fatally:'
+ cat doxygen-log.txt
Treating all remaining warnings as errors
Failing Doxygen test because the following warnings were treated fatally:
warning: Included by graph for 'Exception.h' not generated, too many nodes. Consider increasing DOT_GRAPH_MAX_NODES.
+ echo 'Please fix these warnings.  To run this test locally, use docs/cpp/source/check-doxygen.sh'
Please fix these warnings.  To run this test locally, use docs/cpp/source/check-doxygen.sh
+ rm -f doxygen-log.txt original-doxygen-log.txt
+ exit 1
```
